pkg/daemon/update.go

```diff
@@ -672,13 +680,37 @@ func (dn *Daemon) writeFiles(files []igntypes.File) error {
 				return fmt.Errorf("failed to retrieve file ownership for file %q: %v", file.Path, err)
 			}
 		}
+		if err := createOrigFile(file.Path); err != nil {
+			return err
+		}
 		if err := writeFileAtomically(file.Path, contents.Data, defaultDirectoryPermissions, mode, uid, gid); err != nil {
 			return err
 		}
 	}
 	return nil
 }
 
+func createOrigFile(fpath string) error {
+	f, err := os.Open(fpath)
+	if err != nil {
+		// the file isn't there, no need to back it up
+		// we could check ENOENT only maybe?
+		return nil
+	}
+	if err := os.MkdirAll(filepath.Dir(fpath), defaultDirectoryPermissions); err != nil {
+		return fmt.Errorf("failed to create directory %q: %v", filepath.Dir(fpath), err)
+	}
+	t, err := renameio.TempFile("", fpath+".mcdorig")
+	if err != nil {
+		return err
+	}
+	defer t.Cleanup()
+	if _, err := io.Copy(t, f); err != nil {
```

Comment:
Do we need to check permissions for what was originally used at install time? IE: `rpm -q --dump $PACKAGE | grep $FILE_PATH`? Or do we believe that defaults are sufficient?

Reply:
I believe default are sufficient and umask should kick in if the parent directory is there maybe 🤔 I'll check this out

Reply:
umask doesn't have anything to do with parent directories.

Since we do have some secret files (like the pull secret) we definitely need to be matching the permissions.  Also, some of the files we write need to be executable.

Reply:
Personally I'd just fork off `exec.Command("cp", "-a", "--reflink=auto", fpath, origFileName(fpath))` since among other things that will automatically use reflinks which `io.Copy()` doesn't know about.  (See also [this code](https://gitlab.gnome.org/GNOME/libglnx/blob/8f34033b256da5a8e2cc66faf8e81cfdfb9cdf37/glnx-fdio.c#L712) which I stole from systemd).

Reply:
Uhm ok I'll shell out to cp directly

Reply:
Of course...since we're about to be *replacing* this file just after this, we could simply `rename()` - the problem with that will be a small window where the file doesn't exist, so the more correct thing would be to use [RENAME_EXCHANGE](https://gitlab.gnome.org/GNOME/libglnx/blob/8f34033b256da5a8e2cc66faf8e81cfdfb9cdf37/glnx-fdio.c#L120).

Reply:
Fixed by forking to cp. thanks Colin & Steve